tests-js/vesting/Vesting.js

```diff
@@ -18,114 +18,210 @@ const Vesting = artifacts.require('Vesting');
         await staking.setImplementation(stakingLogic.address);
         staking = await StakingLogic.at(staking.address);
-        
+
         await token.transfer(a2, "1000");
         await token.approve(staking.address, "1000", {from: a2});
-        
+
         kickoffTS = await staking.kickoffTS.call();
     });
-    
+
     describe('constructor', () => {
         it('sets the expected values', async () => {
-
+            await Vesting.new(token.address, staking.address, root, cliff, duration, feeSharingProxy);
         });
-        
-        it('fails if the 0 address is passed as SOV address', async () => {
 
+        it('fails if the 0 address is passed as SOV address', async () => {
+            await expectRevert(Vesting.new(constants.ZERO_ADDRESS, staking.address, root, cliff, duration, feeSharingProxy), "SOV address invalid");
         });
-        
-        it('fails if the 0 address is passed as token owner address', async () => {
 
+        it('fails if the 0 address is passed as token owner address', async () => {
+            await expectRevert(Vesting.new(token.address, staking.address, constants.ZERO_ADDRESS, cliff, duration, feeSharingProxy), "token owner address invalid");
         });
-        
-        it('fails if the 0 address is passed as staking address', async () => {
 
+        it('fails if the 0 address is passed as staking address', async () => {
+            await expectRevert(Vesting.new(token.address, constants.ZERO_ADDRESS, root, cliff, duration, feeSharingProxy), "staking address invalid");
         });
-        
-        it('fails if the vesting duration is bigger than the max staking duration', async () => {
 
+        it('fails if the vesting duration is bigger than the max staking duration', async () => {
+            await expectRevert(Vesting.new(token.address, staking.address, root, cliff, MAX_DURATION.add(new BN(1)), feeSharingProxy), "duration may not exceed the max duration");
         });
-        
+
         it('fails if the vesting duration is shorter than the cliff', async () => {
+            await expectRevert(Vesting.new(token.address, staking.address, root, 100, 99, feeSharingProxy), "duration must be bigger than or equal to the cliff");
+        });
 
+        it('fails if the 0 address is passed as feeSharingProxy address', async () => {
+            await expectRevert(Vesting.new(token.address, staking.address, root, cliff, duration, constants.ZERO_ADDRESS), "feeSharingProxy address invalid");
         });
     });
-    
+
     describe('stakeTokens', () => {
+        let vesting;
         it('should stake 1,000,000 SOV with a duration of 104 weeks and a 26 week cliff', async () => {
+            vesting = await Vesting.new(token.address, staking.address, root, new BN(26 * 7 * 24 * 60 * 60), new BN(104 * 7 * 24 * 60 * 60), feeSharingProxy);
 
+            await token.approve(vesting.address, ONE_MILLON);
+            await vesting.stakeTokens(ONE_MILLON);
```

Comment:
also check that the positions exist on the staking contract and that the voting right is delegated to the token owner

Reply:
this is not implemented in Staking contract.

function setNewStakingContract(address _newStakingContract) public onlyOwner {
        require(_newStakingContract != address(0), "can't reset the new staking contract to 0");
        newStakingContract = _newStakingContract;
    }

function migrateToNewStakingContract() public {
        require(newStakingContract != address(0), "there is no new staking contract set");
        //implementation:
        //iterate over all possible lock dates from now until now + MAX_DURATION
        //read the stake & delegate of the msg.sender
        //if stake > 0, stake it at the new contract until the lock date with the current delegate
    }

Reply:
it should be tested on the Stake contract

Reply:
Done